### PR TITLE
refactor(router-core): slightly faster deepEqual

### PR DESCRIPTION
**Pull request by Sheraff**
## description
- inline `getObjectKeys` as regular control flow inside `deepEqual` to be able to skip as much work as possible, and create fewer objects
- using a regular `for` loop for the array case to skip the overhead of using native array iterators w/ the `.every()` or `.some()` methods
- check `Array.isArray` before `isPlainObject` because this check is cheaper
- `opts.partial` is a special case that allows us to skip a bunch of work, so we give it its own branch

## benchmark

TL;DR: consistently 1.7x faster implementation

```ts
describe.only('deepEqual', () => {
  bench('old implementation', () => {
    deepEqualOld({ a: 1, b: [2, 3] }, { a: 1, b: [2, 3] }, {partial: true})
    deepEqualOld({ a: 1, b: [2, 3] }, { a: 2, b: [2] })
  })

  bench('new implementation', () => {
    deepEqual({ a: 1, b: [2, 3] }, { a: 1, b: [2, 3] }, {partial: true})
    deepEqual({ a: 1, b: [2, 3] }, { a: 2, b: [2] })
  })
})
```

```sh
 ✓  @tanstack/router-core  tests/utils.bench.ts > deepEqual 2484ms
     name                          hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · old implementation  3,806,354.20  0.0002  4.3084  0.0003  0.0003  0.0005  0.0007  0.0013  ±1.78%  1903178
   · new implementation  6,978,154.05  0.0001  0.1937  0.0001  0.0002  0.0002  0.0002  0.0003  ±0.15%  3489078   fastest

 BENCH  Summary

   @tanstack/router-core  new implementation - tests/utils.bench.ts > deepEqual
    1.83x faster than old implementation
```

----

This PR also adds test cases for using `deepEqual` w/ objects containing Symbol and non-enumerable keys. The behavior for those was not changed in this PR, we're just documenting it: they are simply ignored.

----

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved deep-equality checks for arrays and object comparisons; ignore-undefined handling clarified and made more consistent.

* **Refactor**
  * Iteration and comparison logic reorganized for earlier short-circuiting; behavior around enumerable (including inherited) keys adjusted; no public API changes.

* **Tests**
  * Added tests demonstrating edge cases with symbol/non-enumerable properties and prototype augmentation (marked as expected failures).

### CODE REVIEW

File: packages/router-core/src/utils.ts [hunk 1]
```diff
@@ -307,9 +307,12 @@ export function isPlainArray(value: unknown): value is Array<unknown> {
...
 function getObjectKeys(obj: any, ignoreUndefined: boolean) {
-  let keys = Object.keys(obj)
-  if (ignoreUndefined) {
-    keys = keys.filter((key) => obj[key] !== undefined)
+  if (!ignoreUndefined) return Object.keys(obj)
+  const keys = []
+  for (const key in obj) {
+    if (obj[key] !== undefined) {
+      keys.push(key)
+    }
```
[the remark this answers is not in the document]
Comment (Sheraff) [reply packages/router-core/src/utils.ts]: This is is a very smart point, however in this case we're using it behind `isPlainObject` which ensures the objects have either the Object prototype, or the null prototype. So there won't be any inherited enumerable properties (unless the `Object` prototype itself has been modified, but then it's the user's fault for making bad decisions).
